vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+        Some(self)
+    }
+
+    async fn sync_cache(&self) -> Result<(), DiskError> {
+        tracing::trace!("sync_cache");
+
+        (self.conn.lock())
+            .pragma_update(None, "wal_checkpoint", "FULL")
+            .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    async fn unmap(
+        &self,
+        sector_offset: u64,
+        sector_count: u64,
+        _block_level_only: bool,
+        next_is_zero: bool,
+    ) -> Result<(), DiskError> {
+        tracing::trace!(sector_offset, sector_count, "unmap");
+        if sector_offset + sector_count > self.meta.sector_count {
+            return Err(DiskError::IllegalBlock);
+        }
+
+        unblock({
+            let conn = self.conn.clone();
+            move || unmap_sectors(conn, sector_offset, sector_count, next_is_zero)
+        })
+        .await
+        .map_err(|e| DiskError::Io(std::io::Error::new(std::io::ErrorKind::Other, e)))?;
+
+        Ok(())
+    }
+
+    fn unmap_behavior(&self) -> UnmapBehavior {
+        // This layer zeroes if the lower layer is zero, but otherwise does
+        // nothing, so we must report unspecified.
+        UnmapBehavior::Unspecified
+    }
+
+    fn optimal_unmap_sectors(&self) -> u32 {
+        1
+    }
+}
+
+impl WriteNoOverwrite for SqliteDiskLayer {
+    async fn write_no_overwrite(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> Result<(), DiskError> {
+        self.write_maybe_overwrite(buffers, sector, false).await
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+enum SectorKind {
+    AllZero,
+    AllOne,
+    Data(Vec<u8>),
+}
+
+// FUTURE: pass RequestBuffers into this function directly, and avoid the alloc
+// overhead.
+fn read_sectors(
+    conn: Arc<Mutex<Connection>>,
+    sector_size: u32,
+    start_sector: u64,
+    end_sector: u64,
+) -> Result<Vec<(u64, SectorKind)>, rusqlite::Error> {
+    let conn = conn.lock();
+
+    let mut select_stmt = conn.prepare_cached(
+        "SELECT sector, data
+        FROM sectors
+        WHERE sector >= ? AND sector < ?
+        ORDER BY sector ASC",
+    )?;
+    let mut rows = select_stmt.query(rusqlite::params![start_sector, end_sector])?;
+
+    // reuse the same blob handle across all reads.
+    //
+    // this handle isn't easy to persist across multiple calls to
+    // `read_sectors`, due to the lifetime restrictions imposed by using a
+    // `Arc<Mutex<Connection>>`.
+    let Some(first_row) = rows.next()? else {
+        return Ok(Vec::new());
+    };
+    let mut sector: u64 = first_row.get(0)?;
+    let mut blob = conn.blob_open(
+        rusqlite::DatabaseName::Main,
+        "sectors",
+        "data",
+        sector as i64,
+        true,
+    )?;
+
+    let mut buf = vec![0; sector_size as usize];
+    let mut res = Vec::new();
+    loop {
+        let data = match blob.len() {
+            0 => SectorKind::AllZero,
+            1 => SectorKind::AllOne,
```

Comment:
Should we read the blob value to ensure it's not some other sentinel you invent in the future? Or just say that length 1 always means "1s" no matter what the value is?

Reply:
The latter.

FWIW though, I have no hard numbers to suggest if this sort of "all 1s" optimization is even worth it or not... If you think its a bogus optimization, we might be able to do even better by allowing the BLOB to be null, and using that as an indication of all-zero instead?

Reply:
It's not obvious to me that the all-ones optimization is worth it either.